### PR TITLE
Add basic Sourcify verification

### PR DESCRIPTION
**Pull request by marcocastignoli**
Add the basic functions, tasks and tests to provide Sourcify verification

### CODE REVIEW

File: packages/hardhat-verify/src/internal/sourcify.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  address: string;
+  files: {
+    [index: string]: string;
+  };
+  sourceName: string;
+  chosenContract?: number;
+}
+
+// Used for polling the result of the contract verification.
+const VERIFICATION_STATUS_POLLING_TIME = 3000;
```
Comment (schaable) [inline packages/hardhat-verify/src/internal/sourcify.ts:25]:
```suggestion
```

File: packages/hardhat-verify/src/internal/tasks/sourcify.ts [hunk 2]
```diff
@@ -1,28 +1,179 @@
...
+  verify(params: VerificationInterfaceVerifyParams): Promise<any>;
+  getVerificationStatus(guid: string): Promise<any>;
+  getContractUrl(address: string, status?: string): string;
+}
+
+interface AttemptVerificationArgs {
+  address: string;
+  verificationInterface: VerificationInterface;
+  contractFQN: string;
+}
```
Comment (schaable) [inline packages/hardhat-verify/src/internal/tasks/sourcify.ts:49]: The verification interface is Sourcify. Don't redefine it, you can do:
```suggestion

interface AttemptVerificationArgs {
  address: string;
  verificationInterface: Sourcify;
  contractFQN: string;
}
```

File: packages/hardhat-verify/test/integration/index.ts [hunk 3]
```diff
@@ -622,3 +628,52 @@ https://hardhat.etherscan.io/address/${bothLibsContractAddress}#code\n`);
...
+    before(async function () {
+      await this.hre.run(TASK_COMPILE, { force: true, quiet: true });
+      simpleContractAddress = await deployContract(
+        "SimpleContract",
+        [],
+        this.hre
+      );
+    });
+
+    it("should verify a contract on Sourcify", async function () {
```
Comment (schaable) [inline packages/hardhat-verify/test/integration/index.ts:648]: This is ok, but I'd add some tests for the error scenarios as well.

**Review by schaable — CHANGES_REQUESTED** [review]
Hey @marcocastignoli, thanks for sending in the PR! I've dropped a few comments for you. If you want more info or have any questions, feel free to ask.

File: packages/hardhat-verify/src/internal/tasks/sourcify.ts [hunk 4]
```diff
@@ -1,28 +1,179 @@
...
+subtask(TASK_VERIFY_SOURCIFY)
+  .addParam("address")
+  .addFlag("listNetworks")
+  .addParam("contractFQN")
+  .setAction(
+    async (
+      { address, contractFQN }: VerificationArgs,
+      { config, network, run, artifacts }
+    ) => {
+      if (!network.config.chainId) {
```
Comment (schaable) [inline packages/hardhat-verify/src/internal/tasks/sourcify.ts:66]: Sorry for leaving a comment after the initial review. I wasn't sure about this so I wanted to confirm it first. `network.config.chainId` is not reliable to get the chain id, as this value is obtained from the user config. You need to use the `eth_chainId` RPC method, like [here](https://github.com/NomicFoundation/hardhat/blob/1c472a9e6dce792f7a9a94da4db8f3db282efedf/packages/hardhat-verify/src/internal/etherscan.ts#L49).

File: packages/hardhat-verify/src/internal/sourcify.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  private _browserUrl: string;
+  private _chainId: number;
+
+  constructor(chainId: number) {
+    this._apiUrl = "https://sourcify.dev/server";
+    this._browserUrl = "https://repo.sourcify.dev";
+    this._chainId = chainId;
+  }
+
+  // https://docs.sourcify.dev/docs/api/server/check-all-by-addresses/
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply packages/hardhat-verify/src/internal/sourcify.ts:38]: Old docs url, the new one is: https://sourcify.dev/server/api-docs/#/Repository/get_check_all_by_addresses

File: packages/hardhat-verify/src/internal/sourcify.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        _contract.address === address
+    );
+    if (contract.status === "perfect" || contract.status === "partial") {
+      return contract.status;
+    } else {
+      return false;
+    }
+  }
+
+  // https://docs.sourcify.dev/docs/api/server/verify/
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply packages/hardhat-verify/src/internal/sourcify.ts:62]: here again: https://sourcify.dev/server/api-docs/#/Stateless%20Verification/post_verify

File: packages/hardhat-verify/src/internal/sourcify.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        new URL(this._apiUrl),
+        JSON.stringify(parameters)
+      );
+    } catch (error) {
+      throw new ContractVerificationRequestError(this._apiUrl, error as Error);
+    }
+
+    if (!(response.statusCode >= 200 && response.statusCode <= 299)) {
+      // This could be always interpreted as JSON if there were any such guarantee in the Sourcify API.
+      const responseText = await response.body.text();
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply packages/hardhat-verify/src/internal/sourcify.ts:87]: yes but the message key is not always consistent, so I would print the JSON.stringify of the error

File: packages/hardhat-verify/src/internal/sourcify.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    const response = await sendGetRequest(url);
+    const json = await response.body.json();
+
+    const contract = json.find(
+      (_contract: { address: string; status: string }) =>
+        _contract.address === address
+    );
+    if (contract.status === "perfect" || contract.status === "partial") {
+      return contract.status;
```
[the remark this answers is not in the document]
Comment (marcocastignoli) [reply packages/hardhat-verify/src/internal/sourcify.ts:56]: we actually need the status string so we can process the url to the contract on Sourcify

**Pull request merged**